File: src/main/java/eu/mikroskeem/moarsms/Hash.java
package eu.mikroskeem.moarsms;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * @author Mark Vainomaa
 */
final class Hash {
    private Hash() {}

    static String md5(String txt) {
        try {
            // http://stackoverflow.com/a/25251120
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] array = md.digest(txt.getBytes());
            StringBuilder sb = new StringBuilder();
            for (byte anArray : array) {
                sb.append(Integer.toHexString((anArray & 0xFF) | 0x100).substring(1, 3));
            }
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/java/eu/mikroskeem/moarsms/Hash.kt
package eu.mikroskeem.moarsms

import java.security.MessageDigest
import kotlin.experimental.and
import kotlin.experimental.or

/**
 * @author Mark Vainomaa
 */
private const val zero_x_ff: Byte = (0xFF).toByte()
private const val zero_x_hundred: Byte = (0x100).toByte()

internal object Hash {
    internal fun md5(text: String): String {
        val md = MessageDigest.getInstance("MD5")
        val array = md.digest(text.toByteArray())
        val sb = StringBuilder()

        array.forEach {
            sb.append(Integer.toHexString(((it and zero_x_ff) or zero_x_hundred).toInt()).subSequence(1, 3))
        }
        return sb.toString()
    }
}